zkvm/tests/zkvm.rs

```diff
@@ -0,0 +1,81 @@
+use bulletproofs::BulletproofGens;
+use curve25519_dalek::scalar::Scalar;
+use hex;
+use merlin::Transcript;
+
+use zkvm::*;
+
+fn issue_contract(
+    qty: u64,
+    flv: Scalar,
+    issuance_pred: &Predicate,
+    nonce_pred: &Predicate,
+    recipient_pred: &Predicate,
+) -> Vec<Instruction> {
+    vec![
+        Instruction::Push(
+            Commitment::from(CommitmentWitness {
+                value: qty.into(),
+                blinding: Scalar::from(1u64),
+            })
+            .into(),
+        ), // stack: qty
+        Instruction::Var, // stack: qty-var
+        Instruction::Push(Commitment::from(CommitmentWitness::unblinded(flv)).into()), // stack: qty-var, flv
+        Instruction::Var,                                 // stack: qty-var, flv-var
+        Instruction::Push(issuance_pred.clone().into()),  // stack: qty-var, flv-var, pred
+        Instruction::Issue,                               // stack: issue-contract
+        Instruction::Push(nonce_pred.clone().into()),     // stack: issue-contract, pred
+        Instruction::Nonce,                               // stack: issue-contract, nonce-contract
+        Instruction::Signtx,                              // stack: issue-contract
+        Instruction::Signtx,                              // stack: issued-value
+        Instruction::Push(recipient_pred.clone().into()), // stack: issued-value, pred
+        Instruction::Output(1),                           // stack: empty
+    ]
+}
+
+#[test]
+fn issue() {
+    let (tx, txid, txlog) = {
+        // Generate predicates
+        let issuance_pred =
+            Predicate::from_witness(PredicateWitness::Key(Scalar::from(0u64))).unwrap();
+        let nonce_pred =
+            Predicate::from_witness(PredicateWitness::Key(Scalar::from(1u64))).unwrap();
+        let recipient_pred =
+            Predicate::from_witness(PredicateWitness::Key(Scalar::from(2u64))).unwrap();
+
+        // Generate flavor scalar
+        let mut t = Transcript::new(b"ZkVM.issue");
+        t.commit_bytes(b"predicate", issuance_pred.point().as_bytes());
+        let flavor = t.challenge_scalar(b"flavor");
+
+        // Build program
+        let program = issue_contract(1u64, flavor, &issuance_pred, &nonce_pred, &recipient_pred);
```

Comment:
Addition to the above comment: these predicates are not used again after this call, so if we make them _moved_ into the `issue_contract`, then we wouldn't need to clone them.